File: src/java/foundgine-core/src/main/java/com/foundgine/core/execution/ExecutionEvidence.java
package com.foundgine.core.execution;

import java.util.List;

/**
 * Port of {@code Foundgine.Core.Execution.ExecutionEvidence}.
 *
 * <p>Provider-neutral provenance for one execution. Evidence describes what
 * was executed and which authorization boundaries were present without
 * retaining request objects, expression trees, or provider-specific runtime
 * state.
 */
public record ExecutionEvidence(
        String provider,
        String planFingerprint,
        List<Integer> authorizedNodeIds,
        int rowsReturned,
        long elapsedMilliseconds,
        String providerOperationFingerprint,
        String intentFingerprint,
        String authorizationFingerprint,
        String warrantId,
        String warrantDigest,
        String securityInvariantDigest,
        Long authorizationVersion) {

    public ExecutionEvidence(
            String provider,
            String planFingerprint,
            List<Integer> authorizedNodeIds,
            int rowsReturned,
            long elapsedMilliseconds) {
        this(provider, planFingerprint, authorizedNodeIds, rowsReturned, elapsedMilliseconds,
                null, null, null, null, null, null, null);
    }
}

File: src/csharp/Foundgine.Core/Execution/ExecutionEvidence.cs
using System.Security.Cryptography;
using System.Text;

namespace Foundgine.Core.Execution;

/// <summary>
/// Provider-neutral provenance for one execution. Evidence describes what was
/// executed and which authorization boundaries were present without retaining
/// request objects, expression trees, or provider-specific runtime state.
/// </summary>
public sealed record ExecutionEvidence(
    string Provider,
    string PlanFingerprint,
    IReadOnlyList<int> AuthorizedNodeIds,
    int RowsReturned,
    long ElapsedMilliseconds,
    string? ProviderOperationFingerprint = null,
    string? IntentFingerprint = null,
    string? AuthorizationFingerprint = null,
    string? WarrantId = null,
    string? WarrantDigest = null,
    string? SecurityInvariantDigest = null,
    long? AuthorizationVersion = null);

public static class ExecutionEvidenceFactory
{
    public static ExecutionEvidence Create(
        string provider,
        string planFingerprint,
        IEnumerable<int> authorizedNodeIds,
        int rowsReturned,
        long elapsedMilliseconds,
        string? providerOperation = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(provider);
        ArgumentException.ThrowIfNullOrWhiteSpace(planFingerprint);

        return new ExecutionEvidence(
            provider,
            planFingerprint,
            authorizedNodeIds.OrderBy(x => x).ToArray(),
            rowsReturned,
            elapsedMilliseconds,
            providerOperation is null ? null : Hash(providerOperation));
    }

    public static string Hash(string value) =>
        Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(value)));
}